File: projects/02-rate-limiter/src/redis_limiter.rs
//! V3 — Distributed limiter: shared state in Redis, atomic via Lua.
//!
//! V1/V2 are correct for one process. The moment you run **N** gateway instances
//! behind a load balancer, an in-memory bucket per instance enforces `N ×` the
//! intended limit. The fix is to put the state in a shared store (Redis) — but
//! naively that reintroduces a race:
//!
//! ```text
//!   instance A: GET count -> 99      instance B: GET count -> 99
//!   instance A: 99 < 100, allow      instance B: 99 < 100, allow
//!   instance A: SET 100              instance B: SET 100   // both admitted!
//! ```
//!
//! That read-modify-write must be **atomic**. A Lua script runs atomically inside
//! Redis: do the refill math and the deduct server-side, return the decision in
//! one round-trip. Load it once with `SCRIPT LOAD`, then call it by SHA with
//! `EVALSHA` (falling back to `EVAL` on `NOSCRIPT`).

use redis::aio::ConnectionManager;

use crate::error::AppError;
use crate::limiter::{Algorithm, Decision, LimitConfig};

/// Atomic token-bucket update, in Lua, executed inside Redis.
///
/// TODO(V3): write this. Rough shape of a token-bucket script:
///   KEYS[1] = bucket key
///   ARGV    = capacity, refill_per_sec, now_ms, cost, ttl
///   - read stored {tokens, last_refill} (or start full),
///   - refill = min(capacity, tokens + (now-last)/1000 * refill_per_sec),
///   - if refill >= cost: tokens = refill - cost; allowed = 1 else allowed = 0,
///   - HSET the new state, PEXPIRE the key (ttl), return {allowed, remaining, retry_ms}.
/// Keep ALL arithmetic in the script — that's what makes it race-free.
const BUCKET_LUA: &str = r#"
-- TODO(V3): atomic token-bucket refill + acquire.
return redis.error_reply("not implemented")
"#;

/// The production limiter. Cheap to clone (the connection manager is `Arc` inside).
#[derive(Clone)]
pub struct RedisLimiter {
    conn: ConnectionManager,
    cfg: LimitConfig,
    algorithm: Algorithm,
    /// When Redis is unreachable: `true` allows (fail open), `false` denies.
    fail_open: bool,
}

impl RedisLimiter {
    pub fn new(
        conn: ConnectionManager,
        cfg: LimitConfig,
        algorithm: Algorithm,
        fail_open: bool,
    ) -> Self {
        Self {
            conn,
            cfg,
            algorithm,
            fail_open,
        }
    }

    /// Atomically account for a request costing `cost` against `key`.
    pub async fn check(&self, key: &str, cost: u64) -> Result<Decision, AppError> {
        if key.is_empty() {
            return Err(AppError::InvalidArgument("key must not be empty".into()));
        }
        // TODO(V3): run the atomic update in Redis.
        //   1. Pick the script for `self.algorithm` (token bucket here; a
        //      sliding-window variant for `Algorithm::SlidingWindow`).
        //   2. `EVALSHA` it (load once, cache the SHA); on `NOSCRIPT`, `EVAL`
        //      then retry by SHA.
        //   3. Decode the returned {allowed, remaining, retry_ms} into a Decision.
        //   4. On a Redis transport error, honor `self.fail_open`: allow or deny
        //      instead of propagating — and log it.
        let _ = (cost, BUCKET_LUA, &self.conn, self.cfg, self.fail_open);
        todo!("V3: atomic Redis+Lua rate-limit check")
    }

    /// Report current state for `key` WITHOUT consuming budget.
    pub async fn peek(&self, key: &str) -> Result<Decision, AppError> {
        if key.is_empty() {
            return Err(AppError::InvalidArgument("key must not be empty".into()));
        }
        // TODO(V3): read the stored bucket state and report remaining budget
        // without mutating it (a read-only EVAL, or HGETALL + the refill math).
        todo!("V3: peek at Redis bucket state without consuming")
    }
}

File: projects/02-rate-limiter/src/rate_limiter/redis_limiter.py
"""V3 — Distributed limiter: shared state in Redis, made atomic with Lua.

V1 and V2 are correct for one process. Run **N** gateway instances behind a load
balancer and an in-memory bucket per instance enforces `N x` the intended limit —
each one is confidently allowing its own share. The state has to move somewhere
they all see.

Moving it to Redis is not, by itself, the fix. The naive version reintroduces the
race in a new place::

    instance A: HGETALL bucket -> 1 token left    instance B: HGETALL -> 1 token left
    instance A: 1 >= 1, allow                     instance B: 1 >= 1, allow
    instance A: HSET tokens 0                     instance B: HSET tokens 0

Both admitted. Two round-trips means a window between the read and the write, and
that window is all a concurrent instance needs. `INCR` is atomic but can't express
a token bucket — it has no idea what time it is or how fast to refill.

The fix is to stop shipping the decision over the network and ship the *code*
instead: a Lua script that Redis runs **atomically**, doing the refill maths, the
deduct, and the expiry server-side, and returning only the verdict. Redis
executes scripts one at a time, so there is no window to lose.

**Two Python-specific things to get right.**

*Calling the script.* `redis.asyncio.Redis.register_script(source)` returns a
callable that already does the `EVALSHA` -> `NOSCRIPT` -> `EVAL` -> retry dance
for you. Use it in the end, but do the dance by hand at least once
(`script_load`, then `evalsha`, catching `redis.exceptions.NoScriptError`) —
knowing *why* there is a fallback (a Redis restart or `SCRIPT FLUSH` empties the
script cache mid-flight) is the checklist item, not the call.

*Where "now" comes from.* Do not pass `time.monotonic()` in as an argument. N
instances have N slightly different clocks, and a limiter whose refill depends on
whose instance asked is not a shared limiter. Redis has `redis.call("TIME")`,
which returns `{seconds, microseconds}` from the one machine everybody agrees on.
Take the timestamp inside the script. (This is also why scripts must be
deterministic in older Redis — read up on why `TIME` was once forbidden before
replicate-commands, it's a good story about replication.)

Scaffold state: the limiter constructs and holds a live client; the first
`check` raises. That is your worklist.
"""

from __future__ import annotations

import structlog
from redis.asyncio import Redis

from .errors import InvalidArgument
from .limiter import Algorithm, Decision, LimitConfig

__all__ = ["RedisLimiter"]

log = structlog.get_logger(__name__)

# TODO(V3): the atomic token-bucket update, in Lua, executed inside Redis.
#
# Rough shape:
#   KEYS[1] = the bucket key
#   ARGV    = capacity, refill_per_sec, cost, ttl_seconds
#   - now    = redis.call("TIME")            -- {secs, usecs}, Redis's clock
#   - read the stored {tokens, last_refill} (HMGET); a missing key means a fresh,
#     full bucket -- the same "starts full" rule as V1
#   - refilled = math.min(capacity, tokens + (now - last_refill) * refill_per_sec)
#   - if refilled >= cost then tokens = refilled - cost; allowed = 1
#     else allowed = 0; retry = (cost - refilled) / refill_per_sec
#   - HSET the new state, EXPIRE the key (ttl), return {allowed, remaining, retry_ms}
#
# Keep ALL of the arithmetic inside the script. The moment one step happens in
# Python, you have reopened the window this whole vertical exists to close.
#
# Lua notes that will bite you: tables are 1-indexed, so it is `ARGV[1]`;
# everything arriving in ARGV is a *string* (`tonumber` it); and Redis converts
# Lua numbers to integers on the way out, truncating the fraction — so return
# milliseconds, not seconds, or your retry hint is always 0.
BUCKET_LUA = """
-- TODO(V3): atomic token-bucket refill + acquire.
return redis.error_reply("not implemented")
"""

# TODO(V3): the sliding-window-counter variant, for ALGORITHM=sliding_window.
# Same atomicity argument, different arithmetic: two counters and a weight
# instead of a token balance. Note that its natural key layout differs too —
# think about whether the two fixed-window counts want one hash or two keys with
# their own TTLs, and what that does to the "idle keys self-evict" criterion.
WINDOW_LUA = """
-- TODO(V3): atomic sliding-window-counter decision.
return redis.error_reply("not implemented")
"""


class RedisLimiter:
    """The limiter the gRPC service actually calls.

    Holds the shared client rather than a connection: `redis.asyncio.Redis` is a
    pool in front of the socket, so concurrent coroutines each get their own
    connection and a slow key never blocks the others. The pool's size is a
    graded decision — see the SPEC's "bounded pool sized on purpose".
    """

    def __init__(
        self,
        client: Redis,
        config: LimitConfig,
        algorithm: Algorithm,
        *,
        fail_open: bool,
        key_ttl_seconds: int,
    ) -> None:
        self._client = client
        self._config = config
        self._algorithm = algorithm
        # When Redis is unreachable: True allows the request (availability over
        # protection), False denies it (protection over availability). There is
        # no correct answer in general — a login endpoint and a search endpoint
        # want opposite ones — which is why the SPEC grades it as an *explicit*
        # decision rather than a default.
        self._fail_open = fail_open
        self._key_ttl_seconds = key_ttl_seconds

    @property
    def algorithm(self) -> Algorithm:
        return self._algorithm

    @property
    def config(self) -> LimitConfig:
        return self._config

    @property
    def fail_open(self) -> bool:
        return self._fail_open

    def bucket_key(self, key: str) -> str:
        """Namespace a caller's key so this service owns its own keyspace.

        Wired, not a TODO, because a shared Redis with unprefixed keys is how
        one project silently reads another's state.
        """
        return f"ratelimit:{self._algorithm}:{key}"

    async def check(self, key: str, cost: int) -> Decision:
        """Atomically account for a request costing `cost` against `key`."""
        self._validate(key, cost)
        # TODO(V3): run the atomic update in Redis.
        #   1. Pick the script for `self._algorithm` (BUCKET_LUA / WINDOW_LUA).
        #   2. Run it against `self.bucket_key(key)` — by SHA, with the
        #      NOSCRIPT fallback (see the module docstring).
        #   3. Redis hands Lua's return table back as a `list[int]`. Unpack it
        #      into a `Decision` and convert the retry milliseconds to seconds.
        #   4. Catch `redis.exceptions.ConnectionError` / `TimeoutError` and
        #      honour `self._fail_open` — allow or deny rather than letting the
        #      error reach the caller — and log it. Note which exceptions this
        #      should NOT swallow: a `ResponseError` from a bug in your Lua is
        #      not a backend outage, and quietly failing open on it would hide
        #      the bug behind an unlimited endpoint.
        raise NotImplementedError("V3: atomic Redis+Lua rate-limit check")

    async def peek(self, key: str) -> Decision:
        """Report current state for `key` WITHOUT consuming budget."""
        self._validate(key, cost=0)
        # TODO(V3): read the stored state and report the remaining budget
        # without mutating it. A read-only script (the refill maths still has to
        # happen somewhere, and it still needs Redis's clock) is the honest
        # version; `HGETALL` plus the maths in Python is the tempting one — and
        # is wrong for the same TOCTOU reason, just less obviously.
        raise NotImplementedError("V3: peek at Redis bucket state without consuming")

    def _validate(self, key: str, cost: int) -> None:
        """Reject nonsense before it reaches Redis.

        Wired up: the "reject malformed requests" checklist item is about the
        *policy*, and the policy is that garbage never becomes a network
        round-trip. An empty key would collapse every anonymous caller into one
        shared bucket; an absurd cost is either a bug or an attack.
        """
        if not key:
            raise InvalidArgument("key must not be empty")
        if len(key) > MAX_KEY_LEN:
            raise InvalidArgument(f"key must be at most {MAX_KEY_LEN} characters")
        if cost < 0 or cost > self._config.burst:
            raise InvalidArgument(f"cost must be between 0 and the burst ({self._config.burst})")


MAX_KEY_LEN = 512
"""Longest accepted caller key. A key becomes part of a Redis key; keep it boring."""
